File: TOC/1_a.cpp
//Write a program to design the following DFA
//a. To accept the binary string that start with “0” and ending with “1”

#include<iostream>
using namespace std;
int main() {
	cout<<"DFA Accepting Binary Strings Start with 0 and End with 1";
	int state = 0;
	char input[10];
	cout<<"\nEnter the binary string :";
	gets(input);
	
	for(int i=0;i<strlen(input);i++){
		if(input[i]!='0' && input[i]!='1'){
			cout<<"Not a Binary string";
			return 1;
		}
	}
	
	for(int i=0; i<strlen(input); i++) {
		char ch = input[i];
		if(state == 0) {
			if(ch == '0') {
				state = 1;
			} else {
				state = 3;
			}
		} else if(state == 1) {
			if(ch == '0') {
				state = 1;
			} else {
				state = 2;
			}
		} else if(state == 2) {
			if(ch == '0') {
				state = 1;
			} else {
				state = 2;
			}
		} else if(state == 3) {
			state = 3;
		}
		cout<<"Q"<<state<<",";
	}
	if(state == 2) {
		cout<<"Accepted";
	} else {
		cout<<"Rejected";
	}
	return 0;
}

File: 6th sem/Compiler Design/1_a.py
# DFA accepting binary string that starts with '0' and ends with '1'

print("DFA accepting binary strings that start with '0' and end with '1'.")
print("Enter a binary string:  ", end='')
input_string = input().strip()  # removes leading/trailing whitespace

state = 0  # initial state
print('DFA path: Q0', end=' ')
for char in input_string:
    if char not in '01':
        print("Invalid input. Please enter a binary string.")
        exit()
    else:
        if state == 0:
            if char == '0':
                state = 1
            else:
                state = 3
        elif state == 1:
            if char == '0':
                state = 1
            else:
                state = 2
        elif state == 2:
            if char == '0':
                state = 1
            else:
                state = 2
        elif state == 3:
            state = 3
        print(f"Q{state}", end=' ')

if state == 2:
    print("\nAccepted")
else:
    print("\nNot Accepted")
